File: python/demo_myfuncs.py
import math

# Square root function using Newton method
def sqrt(x:float, guess:float, max_iter:int=1000, x_tol:float=1e-8) -> float:
    
    # Check if x is greater than zero
    if x < 0:
        raise ValueError(f"x must be greater than or equal to zero")

    # Loop til newton method has converged 
    for i in range(max_iter):

        # Check if Newton method has converged
        if abs(guess**2 - x) <= x_tol: 
            #print(f"Converge after {i} iterations")           
            return guess
        
        # Perform Newton method
        guess -= (guess**2 - x) / (2*guess)

    print(f"Unable to converge after {i} iterations")
    return guess

# Recursive factorial function 
def factorial(x:int, s:int=1) -> int:

    # Check if x is greater than zero
    if x < 0:
        raise ValueError(f"x must be greater than zero")
    
    # Check if x is an integer
    if type(x) is not int:
        raise ValueError(f"x must be an integer")

    # Recursive factorial 
    if x == 1:
        return s
    elif x == 0:
        return 1
    else:
        return factorial(x - 1, s*x)

# Determines the value of e^x using a taylor series about the nearest integer 
def exponential(x:float, num:int=10) -> float:

    # Find the nearest integer and raise e to power of that int
    x0 = int(round(x))
    e_x0 = 2.7182818284590451**x0
    e_x = 0

    # Perform taylor expansion using factorial function
    for i in range(num):
        e_x += e_x0 * (x - x0)**i / factorial(i)

    return e_x

# Log base e function using Newton method
def log(x:float, guess:float, max_iter=1000, x_tol=1e-8) -> float:

    # Check if x is greater than zero
    if x <= 0:
        raise ValueError(f"x must be greater than zero")

    # Loop until Newton method converges
    for i in range(max_iter):

        # If Newton method converges, return the result
        if abs(exponential(guess) - x) <= x_tol: 
            #print(f"Converge after {i} iterations")           
            return guess
        
        # Perform Newton method
        guess += -1 + x * exponential(-guess)

    print(f"Unable to converge after {i} iterations")
    return guess

num = 10.25

# Outputs
print(f"Sqrt({num}): \tPython = {math.sqrt(num)} \tMy Function = {sqrt(num, 0.5)} \t Relative difference = {abs(math.sqrt(num) - sqrt(num, 0.5))/math.sqrt(num)}")
print(f"factorial({int(num)}): \tPython = {math.factorial(int(num))} \t\tMy Function = {factorial(int(num))} \t\t\t Relative difference = {abs(math.factorial(int(num)) - factorial(int(num)))/math.factorial(int(num))}")
print(f"exp({num}): \tPython = {math.exp(num)} \tMy Function = {exponential(num)} \t Relative difference = {abs(math.exp(num) - exponential(num))/math.exp(num)}")
print(f"log({num}): \tPython = {math.log(num)} \tMy Function = {log(num, 0.5)} \t Relative difference = {abs(math.log(num) - log(num, 0.5))/math.log(num)}")

File: c/old/demo_myfuncs.c
#include <stdio.h>
#include <math.h>

// Square root function using Newton method
double nSqrt(double x, double guess, int maxIter, double xTol){

    // Check if x is greater than 0
    if (x < 0){
        printf("Value Error: x must be greater than or equal to zero");
        return -1;
    }    

    // Loop til newton method has converged or max iterations reached
    for (int i = 0; i < maxIter; i++){

        // Check if Newton method has converged and return answer
        if (fabs(pow(guess,2) - x) <= xTol){
            return guess;
        }

        // Perform Newton method
        guess -= (pow(guess,2) - x) / (2*guess);
    }

    // Fail message and return 0
    printf("unable to converge after %d iterations", maxIter);
    return -1;
}

// Factorial function 
int nFactorial(int x){

    // Check if x is greater than zero
    if (x < 0){
        printf("Value Error: x must be greater than zero");
        return -1;
    }

    // Perform factorial computation
    int s = 1;
    for (int i = 1; i < x; i++){
        s *= i + 1;
    }
    return s;
}

// Determines the value of e^x using a taylor series about the nearest integer 
double nExponential(double x, int num){

    // Find the nearest integer and raise e to power of that int
    int x0 = x;
    double e_x0 = pow(2.7182818284590451,x0);
    double e_x = 0;

    // Perform taylor expansion using factorial function
    for (int i = 0; i < num; i++){
        e_x += e_x0 * pow((x - x0),i) / nFactorial(i);

    }

    return e_x;
}

// Log base e function using Newton method
double nLog(double x, double guess, int max_iter, double x_tol){

    // Check if x is greater than zero
    if (x <= 0){
        printf("Value Error: x must be greater than zero");
        return -1;
    }

    // Loop until Newton method converges
    for (int i = 0; i < max_iter; i++){

        // If Newton method converges, return the result
        if (fabs(nExponential(guess, 10) - x) <= x_tol){
            return guess;
        }
        
        // Perform Newton method
        guess += -1 + x * nExponential(-guess, 10);

    }

    // Fail message and return -1
    printf("Unable to converge after %d iterations", max_iter);
    return -1;
}

int main(){

    // Ask the user for a number
    float number;
    printf("Enter a number: ");
    scanf("%f", &number);  // Read integer input

    // Define variables that will be used to store results from my functions
    double mySqrt, myExp, myLog = 0;
    int myFac = 0;

    // Compute values using my functions
    mySqrt = nSqrt(number, number, 1000, 1e-6);
    myFac  = nFactorial((int) number);
    myExp  = nExponential(number, 10);
    myLog  = nLog(number, number, 1000, 1e-6);

    // Print and compare to math
    printf("Sqrt(%.5f) \t MyFunc = %.5f\n", number, mySqrt);
    printf("Fact(%d) \t MyFunc = %d\n", (int) number, myFac);
    printf("Exp(%.5f) \t MyFunc = %.5f\n", number, myExp);
    printf("Log(%.5f) \t MyFunc = %.5f\n", number, myLog);

    return 0;
}
